core/controllers/classifier_test.py

```diff
@@ -118,6 +122,67 @@ def test_trained_classifier_handler(self):
             classifier_training_jobs[0].status,
             feconf.TRAINING_JOB_STATUS_COMPLETE)
 
+    def test_email_sent_on_failed_job(self):
+
+        class FakeTrainingJob(object):
+            """Fake training class to invoke failed job functions."""
+            def __init__(self):
+                self.status = feconf.TRAINING_JOB_STATUS_FAILED
+
+        def return_fake_training_job(_):
+            return FakeTrainingJob()
+
+        can_send_emails_ctx = self.swap(
+            feconf, 'CAN_SEND_EMAILS', True)
+        can_send_feedback_email_ctx = self.swap(
+            feconf, 'CAN_SEND_FEEDBACK_MESSAGE_EMAILS', True)
+        fail_training_job = self.swap(
+            classifier_services,
+            'get_classifier_training_job_by_id',
+            return_fake_training_job)
+        with can_send_emails_ctx, can_send_feedback_email_ctx:
+            with fail_training_job:
+                # Adding moderator email to admin config page
+                # for sending emails for failed training jobs.
+                self.login(self.ADMIN_EMAIL, is_super_admin=True)
+                config_property = config_domain.Registry.get_config_property(
```

Comment:
I think this and the next command are more like setup before the user session, right? If so, do these calls above line 135 instead.

Reply:
Done.